feat: 공휴일 API Tikxml 응답 파싱 기능 추가 및 관련 테스트 코드 개선 (#12)

Benjamin8282 opened the pull request:
# PR 요약

이번 PR은 공공 API를 활용하여 공휴일 정보를 받아와 앱에 적용하는 기능을 중심으로 다음과 같은 주요 변경 사항을 포함합니다.

## 1. 공휴일 관리 기능 추가
- `HolidayRepository`를 통해 공휴일 데이터를 원격 API에서 가져오고 로컬 DB에 캐싱하는 기능 구현
- 공휴일 관련 DTO(`HolidayDto`, `HolidayItem`) 및 캐시 메타데이터(`HolidayCacheMetadata`) 엔티티 추가
- 공휴일 데이터를 관리하는 DAO (`HolidayDao`, `HolidayCacheMetadataDao`) 추가 및 `AppDatabase`에 반영

## 2. 알람 스케줄링 개선
- `AlarmScheduler` 및 구현체에서 공휴일을 고려하도록 수정
- `HolidayRepository`를 주입받아 공휴일 여부 판단 후 다음 알람 시간 계산에 반영
- `calculateNextAlarmTime` 함수가 `suspend` 함수로 변경되어 비동기 처리 가능

## 3. 저장소 및 테스트 수정
- `RoutineRepositoryImpl`에 `HolidayRepository` 주입 및 공휴일 체크 기능 추가
- `RoutineRepository` 인터페이스의 `isRoutineApplicableForDate` 메서드를 `suspend` 함수로 변경
- `RoutineRepositoryImplTest`에서 공휴일 관련 의존성 목(mock) 처리 및 테스트 보완

## 4. 빌드 설정 및 의존성 업데이트
- Retrofit, TikXML, 로깅 인터셉터 등 API 호출 및 XML 파싱 관련 라이브러리 추가
- Room 스키마 생성을 위한 `ksp` 설정과 공휴일 API 키(`HOLIDAY_API_KEY`) 빌드 변수 추가
- `kotlin-kapt` 플러그인 추가
- `AndroidManifest.xml`에 인터넷 권한 추가

---
## 관련 이슈
Closes #12

Commit:
feat: 공휴일 API 연동을 위한 기본 설정 및 라이브러리 추가 (#12)

- Retrofit, OkHttp 라이브러리 추가
- local.properties를 이용한 API 키 관리 설정
- AndroidManifest.xml에 인터넷 권한 추가

Commit:
feat: 공휴일 API 연동을 위한 네트워크 모듈 추가 (#12)

- HolidayDto, HolidayApiService, HolidayRepository, NetworkModule 생성

Commit:
feat: 공휴일 데이터 연동 및 관련 로직 수정 (#12)

- CalendarViewModel에서 공휴일 정보를 가져와 달력에 표시
- AlarmSchedulerImpl에서 실제 공휴일 정보를 사용하여 알람 설정
- RoutineRepositoryImpl에서 공휴일 루틴 처리 로직 수정

Commit:
feat: 공휴일 정보 캐싱 기능 추가 (#12)

- HolidayItem을 Room 엔티티로 변환
- HolidayDao, HolidayLocalDataSource 생성
- HolidayRepository에 캐싱 로직 추가
- AppDatabase 및 DatabaseModule 업데이트

Commit:
feat: 공휴일 정보 캐싱 및 갱신 로직 구현 (#12)

- HolidayCacheMetadata 엔티티 및 DAO 추가
- AppDatabase 및 DatabaseModule 업데이트
- HolidayRepository에 캐싱 유효성 검사 및 갱신 로직 추가

Commit:
fix: Room 스키마 내보내기 오류 해결 (#12)

- app/build.gradle.kts에 ksp 설정 추가

Commit:
docs: .gitignore 파일에 app/schemas/ 추가 (#12)

- Room 스키마 내보내기 디렉토리 Git 추적 제외

Commit:
fix: AlarmScheduler.kt calculateNextAlarmTime 함수 suspend 키워드 추가 (#12)

- Conflicting overloads 오류 해결을 위해 함수 시그니처 일치

Commit:
refactor: HolidayDto.kt HolidayItem Room 엔티티로 변경 (#12)

- 공휴일 정보 캐싱을 위해 HolidayItem을 Room 엔티티로 변환

Commit:
refactor: HolidayApiService.kt API 경로 및 파라미터 변경 (#12)

- 공공데이터 API 변경에 맞춰 @GET 경로 및 파라미터 수정

Commit:
refactor: NetworkModule.kt BASE_URL 및 OkHttpClient 설정 변경 (#12)

- BASE_URL을 HTTPS로 변경
- OkHttpClient에 HttpLoggingInterceptor 추가

Commit:
feat: RepositoryModule.kt 추가 및 RoutineRepository 바인딩 (#12)

- RoutineRepositoryImpl을 Hilt에 제공하는 RepositoryModule 추가

Commit:
refactor: AppDatabase.kt HolidayCacheMetadata 엔티티 및 DAO 추가 (#12)

- 공휴일 캐싱 메타데이터를 위한 HolidayCacheMetadata 엔티티 추가
- HolidayCacheMetadataDao 추가 및 데이터베이스 버전 업데이트

Commit:
refactor: DatabaseModule.kt HolidayCacheMetadataDao 제공 및 마이그레이션 설정 (#12)

- HolidayCacheMetadataDao를 Hilt에 제공하도록 설정
- Room 데이터베이스 마이그레이션 설정 변경

Commit:
feat: HolidayRepository.kt 공휴일 캐싱 및 갱신 로직 구현 (#12)

- 캐시된 데이터 확인 및 유효성 검사
- 캐시된 데이터가 없거나 유효하지 않을 경우 API 호출 및 캐시 갱신

Commit:
feat: CalendarViewModel 공휴일 정보 연동 (#12)

- HolidayRepository를 통해 공휴일 정보를 가져와 달력에 표시
- 월 변경 시 공휴일 정보 갱신

Commit:
feat: AlarmSchedulerImpl 공휴일 정보 연동 (#12)

- HolidayRepository를 통해 공휴일 정보를 사용하여 알람 설정
- isHoliday 함수 로직 수정

Commit:
feat: 공휴일 정보 월별 캐싱 및 갱신 로직 구현 (#12)

- HolidayCacheMetadata 엔티티에 year, month 필드 추가 및 복합 기본 키 설정
- HolidayCacheMetadataDao에 월별 메타데이터 조회/삭제 메서드 추가
- HolidayDao에 월별 공휴일 삭제 메서드 추가
- HolidayLocalDataSource에 월별 공휴일 삭제 메서드 추가
- HolidayRepository에 월별 캐싱 유효성 검사 및 갱신 로직 구현

Commit:
fix: AppDatabase.kt 데이터베이스 버전 업데이트 (#12)

- HolidayCacheMetadata 엔티티 추가에 따른 DB 버전 5로 업데이트

Commit:
feat: 공휴일 API XML 응답 파싱 기능 추가 (#12)

- Jackson-dataformat-xml 라이브러리 추가
- HolidayDto를 XML 구조에 맞게 재정의
- NetworkModule에서 JacksonConverterFactory 사용하도록 변경

Commit:
Revert "feat: 공휴일 API XML 응답 파싱 기능 추가 (#12)"

This reverts commit e0051695ce7076960edeb4bdbd0681045178f190.

Commit:
feat: 공휴일 API SimpleXML 응답 파싱 기능 추가 (#12)

- converter-simplexml 라이브러리 추가
- HolidayDto를 SimpleXML 어노테이션에 맞게 재정의
- NetworkModule에서 SimpleXmlConverterFactory 사용하도록 변경

Commit:
refactor: XML 파싱 라이브러리 Jackson에서 SimpleXML로 전환 (#12)

- converter-simplexml 라이브러리 사용
- HolidayDto, HolidayRepository, NetworkModule SimpleXML에 맞게 수정

Commit:
feat: 공휴일 API XmlPullParser 응답 파싱 기능 추가 (#12)

- converter-simplexml 제거 및 XmlPullParser 직접 사용
- HolidayDto를 일반 데이터 클래스로 재정의
- NetworkModule에서 컨버터 팩토리 제거
- HolidayRepository에 XmlPullParser 파싱 로직 추가

Commit:
Revert "feat: 공휴일 API XmlPullParser 응답 파싱 기능 추가 (#12)"

This reverts commit f60c87ed2270eb454cad70530912427a37c07040.

Commit:
feat: 공휴일 API Tikxml 응답 파싱 기능 추가 (#12)

- Tikxml 라이브러리 추가
- HolidayDto를 Tikxml 어노테이션에 맞게 재정의
- NetworkModule에서 TikXmlConverterFactory 사용하도록 변경

Commit:
공휴일 정보 파싱 오류 수정 및 TikXml 라이브러리 버전 다운그레이드

- `HolidayDto.kt`: `@Path` 어노테이션 제거
- `HolidayRepository.kt`:
    - 공휴일 캐시 데이터 및 API 응답 처리 시 `response` 객체 제거
    - API 응답에서 `body.items.item` 대신 `body.item`을 사용하도록 수정
- `CalendarViewModel.kt`, `RoutineRepositoryImpl.kt`, `AlarmSchedulerImpl.kt`:
    - 공휴일 정보 조회 시 `response` 객체 제거 및 Nullable 처리 추가
- `NetworkModule.kt`:
    - Retrofit TikXmlConverterFactory 생성 방식 변경 (`createFactory` -> `create`)
- `gradle/libs.versions.toml`: TikXml 관련 라이브러리 버전 0.8.15에서 0.8.13으로 다운그레이드 (0.8.15 에서 annotation 모듈 로드 안되는 문제)

Commit:
chore: KSP에서 kapt로 어노테이션 프로세서 변경

- `app/build.gradle.kts`:
    - `kotlin-kapt` 플러그인 추가
    - `tikxml-processor` 의존성 구성을 `ksp`에서 `kapt`로 변경

Commit:
Fix: 공휴일 정보 파싱 오류 수정

- `HolidayDto.kt`:
    - `retrofit2.http.Path` import 제거
    - `Body.item`을 `Body.items` (Type: `Items`)로 변경
    - `Items` 데이터 클래스 추가 (`item` List 포함)
    - `HolidayItem.isHoliday`를 `HolidayItem.holidayFlag`로 변경
- `HolidayRepository.kt`:
    - 캐시된 공휴일 및 API 응답을 `HolidayDto`로 변환 시 `Body.items`에 `Items` 객체 사용
    - API 응답에서 `apiResponse.body.item` 대신 `apiResponse.body.items.item`을 사용하도록 수정
- `CalendarViewModel.kt`, `RoutineRepositoryImpl.kt`, `AlarmSchedulerImpl.kt`:
    - 공휴일 정보 접근 시 `holidayResponse.body.item` 대신 `holidayResponse.body.items.item` 사용
    - 공휴일 플래그 확인 시 `it.isHoliday` 대신 `it.holidayFlag` 사용
- `AppDatabase.kt`: 데이터베이스 버전 5에서 6으로 변경

Commit:
refactor: HolidayRepository 및 관련 테스트 코드 수정 (#12)

- HolidayRepository 클래스와 getHolidayInfo 메서드를 open으로 변경하여 테스트 용이성 확보
- RoutineRepositoryImplTest (androidTest)에서 HolidayRepository 의존성 주입 방식 수정
- HolidayRepository의 의존성(HolidayApiService, HolidayLocalDataSource, HolidayCacheMetadataDao)을 목킹하여 테스트 환경 구성

Commit:
test: RoutineRepositoryImplTest 목 객체 추가 및 주석 수정

- `HolidayRepository` 목 객체를 추가하여 테스트 환경을 구성합니다.
- 테스트 클래스 및 `setup` 함수의 주석을 수정하여 `HolidayRepository` 목 객체 사용을 명시합니다.
- `getRoutineItemsForPeriod` 테스트에서 불필요한 `println` 문과 주석을 제거합니다.

f-lab-Toru commented on app/src/main/java/com/example/myroutine/data/alarm/AlarmSchedulerImpl.kt:
```diff
@@ -14,18 +14,22 @@ import java.time.LocalTime
+        routine.alarmTime ?: return
 
-        val nextTriggerTime = calculateNextAlarmTime(routine) ?: return
+        val nextTriggerTime = runBlocking { calculateNextAlarmTime(routine) } ?: return
```
runBlocking{} 은 GlobalScope와 함께 사용을 가급적 피하는 것이 좋은 API 입니다.
withContext 로 리팩토링해도 되지 않을까 싶어요

f-lab-Toru commented on app/src/main/java/com/example/myroutine/data/repository/HolidayRepository.kt:
```diff
@@ -0,0 +1,108 @@
+import kotlin.collections.isNotEmpty
+
+
+open class HolidayRepository @Inject constructor(
```
엄밀히 말하면 HolidayRepository 도 interface로 만들고 구현체로 구현해 주는 것이 flexibility와  scalability 를 위해서 더 좋을 겁니다

f-lab-Toru commented on app/src/main/java/com/example/myroutine/features/calendar/CalendarViewModel.kt:
```diff
@@ -54,9 +55,27 @@ class CalendarViewModel @Inject constructor(
     }
 
-    
+    private fun fetchHolidays(year: Int, month: Int) {
```
이 메서드는 viewModelScope 안에서 그냥 돌도록 하면 되지 않을까 싶네요. 앞에 suspend 만 붙여 주면 되지 않을까요?

f-lab-Toru commented on app/src/main/java/com/example/myroutine/data/repository/HolidayRepository.kt:
```diff
@@ -0,0 +1,108 @@
+        val monthString = month.toString().padStart(2, '0')
+        L.d(TAG, "Calling API with solYear=$year, solMonth=$monthString")
+
+        val apiResponse = try {
```
나중에 API request 를 담당하는 부분을 떼어내서 결과값만 Repository 로 넘기도록 해도 될 거 같습니다

f-lab-Toru commented on app/src/main/java/com/example/myroutine/data/repository/HolidayRepository.kt:
```diff
@@ -0,0 +1,108 @@
+        L.d(TAG, "Loaded lastCachedMetadata: $lastCachedMetadata")
+
+        val isCacheValid = lastCachedMetadata != null &&
+                (System.currentTimeMillis() - lastCachedMetadata.lastCachedTimestamp) < CACHE_EXPIRATION_DAYS * 24 * 60 * 60 * 1000L
```
뒤에 캐시연한초과와 같은 식들은 companion object 안으로 넣어도 되지 않을까요?

f-lab-Toru reviewed (commented):
Repository 에 관한 코파일럿의 리뷰도 참고하면 좋을 거 같습니다

Commit:
refactor: 공휴일 정보 캐싱 기능 추가

- `RoutineRepositoryImpl.kt`:
    - `holidayCache` 추가: 조회한 공휴일 정보를 월별로 캐싱하여 API 호출 최소화
    - `isHoliday` 메소드 수정: 캐시된 공휴일 정보 우선 사용

Commit:
refactor: AlarmScheduler.schedule를 suspend 함수로 변경

- `AlarmScheduler.kt`: `schedule` 함수를 `suspend`로 변경
- `AlarmSchedulerImpl.kt`:
    - `schedule` 함수를 `suspend`로 변경
    - `calculateNextAlarmTime` 호출 시 `runBlocking` 제거

Commit:
refactor: CalendarViewModel holiday fetching 개선

- `fetchHolidays` 함수를 suspend 함수로 변경
- ViewModel 초기화 시 `viewModelScope.launch` 블록 내에서 `fetchHolidays` 호출하도록 수정

Commit:
refactor: 공휴일 캐시 만료 시간 상수화

- `CACHE_EXPIRATION_DAYS`를 기반으로 `CACHE_EXPIRATION_MILLISECONDS` 상수 추가
- 캐시 유효성 검사 시 `CACHE_EXPIRATION_MILLISECONDS` 사용
